File: DirectShowFilters/MPUrlSourceSplitter/MPUrlSourceSplitter/MPUrlSourceSplitter_Protocol_Rtsp/RtspRequest.cpp
/*
    Copyright (C) 2007-2010 Team MediaPortal
    http://www.team-mediaportal.com

    This file is part of MediaPortal 2

    MediaPortal 2 is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    MediaPortal 2 is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with MediaPortal 2.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "StdAfx.h"

#include "RtspRequest.h"
#include "RtspSequenceRequestHeader.h"
#include "RtspSessionRequestHeader.h"

CRtspRequest::CRtspRequest(HRESULT *result)
{
  this->uri = NULL;
  this->request = NULL;
  this->version = NULL;
  this->requestHeaders = NULL;
  this->timeout = 0;

  if ((result != NULL) && (SUCCEEDED(*result)))
  {
    this->requestHeaders = new CRtspRequestHeaderCollection(result);
    CHECK_POINTER_HRESULT(*result, this->requestHeaders, *result, E_OUTOFMEMORY);

    if (SUCCEEDED(*result) && (this->requestHeaders != NULL))
    {
      CRtspSequenceRequestHeader *header = new CRtspSequenceRequestHeader(result);
      CHECK_POINTER_HRESULT(*result, header, *result, E_OUTOFMEMORY);

      CHECK_CONDITION_HRESULT(*result, this->requestHeaders->Add(header), *result, E_OUTOFMEMORY);

      CHECK_CONDITION_EXECUTE(FAILED(*result), FREE_MEM_CLASS(header));
    }
  }
}

CRtspRequest::CRtspRequest(HRESULT *result, bool createDefaultHeaders)
{
  this->uri = NULL;
  this->request = NULL;
  this->version = NULL;
  this->requestHeaders = NULL;
  this->timeout = 0;

  if ((result != NULL) && (SUCCEEDED(*result)))
  {
    this->requestHeaders = new CRtspRequestHeaderCollection(result);
    CHECK_POINTER_HRESULT(*result, this->requestHeaders, *result, E_OUTOFMEMORY);

    if (SUCCEEDED(*result) && createDefaultHeaders && (this->requestHeaders != NULL))
    {
      CRtspSequenceRequestHeader *header = new CRtspSequenceRequestHeader(result);
      CHECK_POINTER_HRESULT(*result, header, *result, E_OUTOFMEMORY);

      CHECK_CONDITION_HRESULT(*result, this->requestHeaders->Add(header), *result, E_OUTOFMEMORY);

      CHECK_CONDITION_EXECUTE(FAILED(*result), FREE_MEM_CLASS(header));
    }
  }
}

CRtspRequest::~CRtspRequest(void)
{
  FREE_MEM(this->uri);
  FREE_MEM(this->request);
  FREE_MEM(this->version);
  FREE_MEM_CLASS(this->requestHeaders);
}

/* get methods */

const wchar_t *CRtspRequest::GetUri(void)
{
  return this->uri;
}

const wchar_t *CRtspRequest::GetVersion(void)
{
  return (this->version == NULL) ? RTSP_REQUEST_VERSION : this->version;
}

const wchar_t *CRtspRequest::GetRequest(void)
{
  bool result = (this->GetRequestHeaders() != NULL);
  FREE_MEM(this->request);

  if (result)
  {
    wchar_t *requestLine = FormatString(RTSP_REQUEST_LINE_FORMAT, this->GetMethod(), this->GetUri(), this->GetVersion(), RTSP_CRLF);
    result &= (requestLine != NULL);

    wchar_t *requestHeader = Duplicate(L"\r\n");
    result &= (requestHeader != NULL);

    for (unsigned int i = 0; (result && (i < this->GetRequestHeaders()->Count())); i++)
    {
      CRtspRequestHeader *header = this->GetRequestHeaders()->GetItem(i);

      wchar_t *temp = FormatString(L"%s%s", header->GetRequestHeader(), requestHeader);
      result &= (temp != NULL);
      FREE_MEM(requestHeader);
      requestHeader = temp;
    }

    if (result)
    {
      this->request = FormatString(RTSP_REQUEST_FORMAT, requestLine, requestHeader);
      result &= (this->request != NULL);
    }

    FREE_MEM(requestLine);
    FREE_MEM(requestHeader);
  }

  return this->request;
}

CRtspRequestHeaderCollection *CRtspRequest::GetRequestHeaders(void)
{
  return this->requestHeaders;
}

unsigned int CRtspRequest::GetSequenceNumber(void)
{
  unsigned int result = RTSP_SEQUENCE_NUMBER_UNSPECIFIED;

  if (this->requestHeaders != NULL)
  {
    CRtspSequenceRequestHeader *header = dynamic_cast<CRtspSequenceRequestHeader *>(this->requestHeaders->GetRtspHeader(RTSP_SEQUENCE_REQUEST_HEADER_NAME, false));
    if (header != NULL)
    {
      result = header->GetSequenceNumber();
    }
  }

  return result;
}

const wchar_t *CRtspRequest::GetSessionId(void)
{
  const wchar_t *result = NULL;

  if (this->requestHeaders != NULL)
  {
    CRtspSessionRequestHeader *header = dynamic_cast<CRtspSessionRequestHeader *>(this->requestHeaders->GetRtspHeader(RTSP_SESSION_REQUEST_HEADER_NAME, false));
    if (header != NULL)
    {
      result = header->GetSessionId();
    }
  }

  return result;
}

unsigned int CRtspRequest::GetTimeout(void)
{
  return this->timeout;
}

/* set methods */

bool CRtspRequest::SetUri(const wchar_t *uri)
{
  SET_STRING_RETURN_WITH_NULL(this->uri, uri);
}

bool CRtspRequest::SetVersion(const wchar_t *version)
{
  SET_STRING_RETURN_WITH_NULL(this->version, version);
}

void CRtspRequest::SetTimeout(unsigned int timeout)
{
  this->timeout = timeout;
}

void CRtspRequest::SetSequenceNumber(unsigned int sequenceNumber)
{
  if (this->requestHeaders != NULL)
  {
    CRtspSequenceRequestHeader *header = dynamic_cast<CRtspSequenceRequestHeader *>(this->requestHeaders->GetRtspHeader(RTSP_SEQUENCE_REQUEST_HEADER_NAME, false));
    if (header != NULL)
    {
      header->SetSequenceNumber(sequenceNumber);
    }
  }
}

bool CRtspRequest::SetSessionId(const wchar_t *sessionId)
{
  HRESULT result = (this->requestHeaders != NULL) ? S_OK : E_NOT_VALID_STATE;

  if (SUCCEEDED(result))
  {
    unsigned int index = this->requestHeaders->GetRtspHeaderIndex(RTSP_SESSION_REQUEST_HEADER_NAME, false);

    if (index != UINT_MAX)
    {
      if (sessionId == NULL)
      {
        CHECK_CONDITION_HRESULT(result, this->requestHeaders->Remove(index), result, E_FAIL);
      }
      else
      {
        CRtspSessionRequestHeader *header = dynamic_cast<CRtspSessionRequestHeader *>(this->requestHeaders->GetItem(index));
        CHECK_POINTER_HRESULT(result, header, result, E_FAIL);

        CHECK_CONDITION_HRESULT(result, header->SetSessionId(sessionId), result, E_OUTOFMEMORY);
      }
    }
    else if (sessionId != NULL)
    {
      // create new session ID header
      CRtspSessionRequestHeader *header = new CRtspSessionRequestHeader(&result);
      CHECK_POINTER_HRESULT(result, header, result, E_OUTOFMEMORY);

      CHECK_CONDITION_HRESULT(result, header->SetSessionId(sessionId), result, E_OUTOFMEMORY);
      CHECK_CONDITION_HRESULT(result, this->requestHeaders->Add(header), result, E_OUTOFMEMORY);
      CHECK_CONDITION_EXECUTE(FAILED(result), FREE_MEM_CLASS(header));
    }
  }

  return SUCCEEDED(result);
}

/* other methods */

//bool CRtspRequest::CreateRequest(void)
//{
//  bool result = (this->GetRequestHeaders() != NULL);
//  FREE_MEM(this->request);
//
//  if (result)
//  {
//    wchar_t *requestLine = FormatString(RTSP_REQUEST_LINE_FORMAT, this->GetMethod(), this->GetUri(), this->GetVersion(), RTSP_CRLF);
//    result &= (requestLine != NULL);
//
//    wchar_t *requestHeader = Duplicate(L"\r\n");
//    result &= (requestHeader != NULL);
//
//    for (unsigned int i = 0; (result && (i < this->GetRequestHeaders()->Count())); i++)
//    {
//      CRtspRequestHeader *header = this->GetRequestHeaders()->GetItem(i);
//
//      wchar_t *temp = FormatString(L"%s%s", header->GetRequestHeader(), requestHeader);
//      result &= (temp != NULL);
//      FREE_MEM(requestHeader);
//      requestHeader = temp;
//    }
//
//    if (result)
//    {
//      this->request = FormatString(RTSP_REQUEST_FORMAT, requestLine, requestHeader);
//      result &= (this->request != NULL);
//    }
//
//    FREE_MEM(requestLine);
//    FREE_MEM(requestHeader);
//  }
//
//  return result;
//}

CRtspRequest *CRtspRequest::Clone(void)
{
  CRtspRequest *result = this->CreateRequest();
  if (result != NULL)
  {
    if (!this->CloneInternal(result))
    {
      FREE_MEM_CLASS(result);
    }
  }
  return result;
}

/* protected methods */

bool CRtspRequest::CloneInternal(CRtspRequest *clone)
{
  bool result = true;

  SET_STRING_AND_RESULT_WITH_NULL(clone->uri, this->uri, result);
  SET_STRING_AND_RESULT_WITH_NULL(clone->version, this->version, result);
  SET_STRING_AND_RESULT_WITH_NULL(clone->request, this->request, result);
  clone->timeout = this->timeout;

  result &= clone->requestHeaders->Append(this->requestHeaders);

  return result;
}

File: TvEngine3/Mediaportal/TV/Server/TVLibrary/Implementations/Rtsp/RtspRequest.cs
#region Copyright (C) 2005-2011 Team MediaPortal

// Copyright (C) 2005-2011 Team MediaPortal
// http://www.team-mediaportal.com
// 
// MediaPortal is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 2 of the License, or
// (at your option) any later version.
// 
// MediaPortal is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with MediaPortal. If not, see <http://www.gnu.org/licenses/>.

#endregion

using System.Collections.Generic;
using System.Text;

namespace Mediaportal.TV.Server.TVLibrary.Implementations.Rtsp
{
  /// <summary>
  /// A simple class that can be used to serialise RTSP requests.
  /// </summary>
  internal class RtspRequest
  {
    private RtspMethod _method = null;
    private string _uri = null;
    private int _majorVersion = 1;
    private int _minorVersion = 0;
    private IDictionary<string, string> _headers = new Dictionary<string, string>();
    private string _body = string.Empty;

    /// <summary>
    /// Initialise a new instance of the <see cref="RtspRequest"/> class.
    /// </summary>
    /// <param name="method">The request method.</param>
    /// <param name="url">The request URI</param>
    /// <param name="majorVersion">The major version number.</param>
    /// <param name="minorVersion">The minor version number.</param>
    public RtspRequest(RtspMethod method, string uri, int majorVersion = 1, int minorVersion = 0)
    {
      _method = method;
      _uri = uri;
      _majorVersion = majorVersion;
      _minorVersion = minorVersion;
    }

    /// <summary>
    /// Get the request method.
    /// </summary>
    public RtspMethod Method
    {
      get
      {
        return _method;
      }
    }

    /// <summary>
    /// Get the request URI.
    /// </summary>
    public string Uri
    {
      get
      {
        return _uri;
      }
    }

    /// <summary>
    /// Get the request major version number.
    /// </summary>
    public int MajorVersion
    {
      get
      {
        return _majorVersion;
      }
    }

    /// <summary>
    /// Get the request minor version number.
    /// </summary>
    public int MinorVersion
    {
      get
      {
        return _minorVersion;
      }
    }

    /// <summary>
    /// Get or set the request headers.
    /// </summary>
    public IDictionary<string, string> Headers
    {
      get
      {
        return _headers;
      }
      set
      {
        _headers = value;
      }
    }

    /// <summary>
    /// Get or set the request body.
    /// </summary>
    public string Body
    {
      get
      {
        return _body;
      }
      set
      {
        _body = value;
      }
    }

    /// <summary>
    /// Serialise this request.
    /// </summary>
    /// <returns>raw request bytes</returns>
    public byte[] Serialise()
    {
      StringBuilder request = new StringBuilder();
      request.AppendFormat("{0} {1} RTSP/{2}.{3}\r\n", _method, _uri, _majorVersion, _minorVersion);
      foreach (KeyValuePair<string, string> header in _headers)
      {
        request.AppendFormat("{0}: {1}\r\n", header.Key, header.Value);
      }
      request.AppendFormat("\r\n{0}", _body);
      return System.Text.Encoding.UTF8.GetBytes(request.ToString());
    }
  }
}
